src/java/org/apache/poi/ss/usermodel/helpers/ColumnShifter.java

```diff
@@ -119,4 +122,87 @@ private boolean removalNeeded(CellRangeAddress merged, int startColumn, int endC
         // if the merged-region and the overwritten area intersect, we need to remove it
         return merged.intersects(overwrite);
     }
+    public void shiftColumns(int firstShiftColumnIndex, int lastShiftColumnIndex, int step){
+        if(step > 0)
+            shiftColumnsRight(firstShiftColumnIndex, lastShiftColumnIndex, step);
+        else if(step < 0)
+            shiftColumnsLeft(firstShiftColumnIndex, lastShiftColumnIndex, -step);
+    }
+    /**
+     * Inserts shiftStep empty columns at firstShiftColumnIndex-th position, and shifts rest columns to the right 
+     * (see constructor for parameters)
+     */
+
+    private void shiftColumnsRight(int firstShiftColumnIndex, int lastShiftColumnIndex, int step){
+        for (Row row : sheet)
+        {    
+            if(row != null){
+                for (int columnIndex = lastShiftColumnIndex; columnIndex >= firstShiftColumnIndex; columnIndex--){ // process cells backwards, because of shifting 
+                    Cell oldCell = row.getCell(columnIndex);
+                    Cell newCell = row.createCell(columnIndex + step);
+                    if(oldCell == null)
+                        nullifyCell(newCell);
+                    else cloneCellValue(oldCell,newCell);
+                }
+                for (int columnIndex = firstShiftColumnIndex; columnIndex <= firstShiftColumnIndex+step-1; columnIndex++)
+                    nullifyCell(row.getCell(columnIndex));
+            }
+        }
+    }
+
+    private void shiftColumnsLeft(int firstShiftColumnIndex, int lastShiftColumnIndex, int step){
+        for (Row row : sheet)
+        {    
+            if(row != null){
+                for (int columnIndex = firstShiftColumnIndex; columnIndex <= lastShiftColumnIndex; columnIndex++){ 
+                    Cell oldCell = row.getCell(columnIndex);
+                    Cell newCell = null;
+                    if(columnIndex - step >= 0){
+                        newCell = row.getCell(columnIndex - step);
+                        if(newCell == null)
+                            newCell = row.createCell(columnIndex - step);
+                        if(oldCell != null)
+                           cloneCellValue(oldCell, newCell);
+                        else nullifyCell(newCell);
+                    }
+                    else throw new IllegalStateException("Column index less than zero : " + (Integer.valueOf(columnIndex - step)).toString());
+                }
+                for (int columnIndex = lastShiftColumnIndex-step+1; columnIndex <= lastShiftColumnIndex; columnIndex++)
+                    nullifyCell(row.getCell(columnIndex));
+                }
+        }
+    }
+    
+    private void nullifyCell(Cell cell){
```

Comment:
I think `Row.removeCell(Cell)` will accomplish what you want here. It will completely remove a cell in the sheet. This includes value, style, borders, etc.

But I don't think we need to remove a Cell, just move the cell to a different column in the Row. It looks like the Row and Cell API don't have methods to move a cell from one column to another. Could you propose an API that would allow you to move a cell to a new column within a row so that you do not have to copy the cell and delete the original?

Reply:
Meanwhile I have refactored shifting, and I have resolved this in the other way. You can see it in new commit.